feat: Refactor rfq comparison price integration

PirosB3 opened the pull request:
## Description

<!--- Describe your changes in detail -->

Calculates a "price to beat" that can be funneled to RFQ makers. For now, this is naively the price of the overall order. Adds an extra `comparisonPrice` parameter to the RFQ GET request.

## Testing instructions

- Added an end-to-end unit test
- Will perform testing locally

<!--- Please describe how reviewers can test your changes -->

## Types of changes

<!--- What types of changes does your code introduce? Uncomment all the bullets that apply: -->

<!-- * Bug fix (non-breaking change which fixes an issue) -->

<!-- * New feature (non-breaking change which adds functionality) -->

<!-- * Breaking change (fix or feature that would cause existing functionality to change) -->

## Checklist:

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

-   [X] Prefix PR title with `[WIP]` if necessary.
-   [X] Add tests to cover changes as needed.
-   [X] Update documentation as needed.
-   [X] Add new entries to the relevant CHANGELOG.jsons.

Commit:
initial implementation

Commit:
refactor market side ops

Commit:
refactor coding and avoid running indicative quotes when sampling market liquidity

Commit:
added initial RFQ tests

Commit:
added more unit tests

Commit:
added further unit tests

Commit:
added a few fixed and added some comments:

Commit:
Merge branch 'development' of github.com:0xProject/0x-monorepo into refactor_rfq_request_positioning

Commit:
added linting and prettifying

Commit:
Merge branch 'development' of github.com:0xProject/0x-monorepo into refactor_rfq_request_positioning

# Conflicts:
#	packages/asset-swapper/src/utils/market_operation_utils/index.ts

Commit:
added initial implementation

Commit:
Added initial unit tests and implementation

PirosB3 commented:
NOTE: this PR will never build until https://github.com/0xProject/quote-server/pull/8 is merged and released

Commit:
Added fix for decimals

Commit:
added fix

Commit:
Mark decimal places

feuGeneA commented on packages/asset-swapper/src/utils/market_operation_utils/index.ts:
```diff
@@ -57,6 +59,7 @@ export async function getRfqtIndicativeQuotesAsync(
+    comparisonPrice: BigNumber | undefined,
     opts: Partial<GetMarketOrdersOpts>,
 ): Promise<RFQTIndicativeQuote[]> {
     if (opts.rfqt && opts.rfqt.isIndicative === true && opts.rfqt.quoteRequestor) {
```
I'm wondering whether this new `comparisonPrice` shouldn't go in the `opts.rfqt` object here rather than getting its own parameter...IF it's only ever intended to be used for RFQT.  If it became part of that object, I believe it would eliminate a lot of the parameter additions you had to do in `quote_requestor`.  Also it would allow you to avoid clients needing to pass in `undefined` for this parameter if they aren't actually using it (like you had to do in your tests), as you could mark the property `?` (optional) in the type definition in the RFQ options object.

PirosB3 replied:
So you are suggesting adding this parameter as part of `GetMarketOrdersOpts`?

feuGeneA commented on packages/asset-swapper/src/utils/market_operation_utils/index.ts:
```diff
@@ -460,7 +472,7 @@ export class MarketOperationUtils {
 
     public async _generateOptimizedOrdersAsync(
-        marketSideLiquidity: MarketSideLiquidity,
+        marketSideLiquidity: Omit<MarketSideLiquidity, 'makerTokenDecimals' | 'takerTokenDecimals'>,
```
Why was it necessary to explicitly omit these properties rather than just letting them come along for the ride?

feuGeneA commented on packages/asset-swapper/src/utils/market_operation_utils/multihop_utils.ts:
```diff
@@ -34,7 +36,7 @@ export function getIntermediateTokens(
  */
 export function getBestTwoHopQuote(
-    marketSideLiquidity: MarketSideLiquidity,
+    marketSideLiquidity: Omit<MarketSideLiquidity, 'makerTokenDecimals' | 'takerTokenDecimals'>,
```
Same as another comment, wondering why these needed to be omitted here.

PirosB3 replied:
I think this was done for testing purposes. In general, bunny-hop does not care about decimals because it simply works with selling + buying base unit amount. Whatever it gets back, it then sells.

feuGeneA commented on packages/asset-swapper/src/utils/quote_requestor.ts:
```diff
@@ -1,6 +1,7 @@
 import { schemas, SchemaValidator } from '@0x/json-schemas';
 import { assetDataUtils, orderCalculationUtils, SignedOrder } from '@0x/order-utils';
 import { RFQTFirmQuote, RFQTIndicativeQuote, TakerRequest } from '@0x/quote-server';
+import { TakerRequestQueryParams } from '@0x/quote-server/lib/src/types';
```
I think you want to export this from the `index.ts` file in `quote-server` (see my comment on that PR), and then here drop the `lib/src/types` suffix... no?  I may be wrong, but my understanding is that referring to package sub-paths is an anti-pattern.

feuGeneA commented on packages/asset-swapper/src/utils/quote_requestor.ts:
```diff
@@ -113,6 +114,39 @@ export class QuoteRequestor {
+        makerAssetData: string,
+        takerAssetData: string,
+        assetFillAmount: BigNumber,
+        comparisonPrice?: BigNumber | undefined,
```
doesn't using `?` imply `|undefined`? (I think you can drop the `|undefined`.)

feuGeneA replied:
Addressed in https://github.com/0xProject/0x-monorepo/pull/2720/commits/2cf31f05a1b402f5e8f7853c60dc760baa0c6b3b

feuGeneA commented on packages/asset-swapper/src/utils/quote_requestor.ts:
```diff
@@ -113,6 +114,39 @@ export class QuoteRequestor {
+    public static makeQueryParameters(
+        takerAddress: string,
+        marketOperation: MarketOperation,
+        makerAssetData: string,
```
Consider merging `makeQueryParameters()` with the existing `inferQueryParameters()` function (using whatever name feels appropriate to the unified concept).  Seems like it would be confusing to come into this code and see both of those functions and try to understand the difference between the two.

PirosB3 replied:
great suggestion! I refactored

feuGeneA reviewed (changes requested):
I'll need to defer to someone else on the Solidity, and also on the suggested price calculation.

Bravo on those typemoq tests! They look like they were a beast to pin down!

Commit:
Merge branch 'development' of github.com:0xProject/0x-monorepo into refactor_rfq_comparison_price_integration

# Conflicts:
#	packages/asset-swapper/src/utils/market_operation_utils/index.ts

Commit:
fixes a few issues with the typing

Commit:
update quote server

Commit:
fixed broken package

Commit:
Merge branch 'development' of github.com:0xProject/0x-monorepo into refactor_rfq_comparison_price_integration

feuGeneA committed:
Merge branch 'development' into refactor_rfq_comparison_price_integration

feuGeneA committed:
Fix linter errors

feuGeneA committed:
Merge branch 'development' into refactor_rfq_comparison_price_integration

feuGeneA committed:
Stop referring to GetMarketOrdersOpts.shouldBat...

Because it doesn't exist on that type.

feuGeneA committed:
Fixed linter error

ERROR: 793:24  no-unnecessary-type-assertion  This assertion is unnecessary since it does not change the type of the expression.

feuGeneA committed:
Ran prettier

feuGeneA committed:
Fix failing test-doc-generation

feuGeneA committed:
Drop `|undefined` from type of optional param

Addresses review comment https://github.com/0xProject/0x-monorepo/pull/2720#discussion_r500453526

feuGeneA committed:
Reduce solidity compiler optimizer runs...

...in order to accomodate now-too-big contract TestERC20BridgeSampler,
which imports the recently-added-to NativeOrderSampler.sol.

Fixes test failing like:
  1) erc20-bridge-sampler
       "before all" hook in "erc20-bridge-sampler":
     RuntimeError: VM Exception while processing transaction: out of gas

Commit:
apply PR feedback

Commit:
cheap lint fix

Commit:
Merge branch 'development' of github.com:0xProject/0x-monorepo into refactor_rfq_comparison_price_integration

# Conflicts:
#	packages/asset-swapper/src/utils/market_operation_utils/index.ts

Commit:
Merge branch 'development' of github.com:0xProject/0x-monorepo into refactor_rfq_comparison_price_integration

Commit:
Add changes for the feature flag

Commit:
comments and renaming

Commit:
small lint fix

feuGeneA committed:
Ran prettier

feuGeneA committed:
Merge branch 'refactor_rfq_comparison_price_integration' into rfqt_comparison_feature_flag

feuGeneA committed:
Disable tests per IS_PRICE_AWARE_RFQ_ENABLED flag

feuGeneA committed:
Ran prettier

feuGeneA committed:
Leave link to PR in comments

PirosB3 committed:
Merge pull request #2735 from 0xProject/rfqt_comparison_feature_flag

Rfqt comparison feature flag
